lib/secretsscanner/scan/scan.go

```diff
@@ -0,0 +1,252 @@
+// and checks if they are SSH private keys.
+func (s *Scanner) findPrivateKeys(ctx context.Context, root, deviceID string) {
+	logger := s.log.With("dir", root)
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		switch fileData, isKey, err := readFileIfSSHPrivateKey(path); {
+		case err != nil:
+			logger.DebugContext(ctx, "error reading file", "path", path, "error", err)
+		case isKey:
+			key, err := extractSSHKey(ctx, path, deviceID, fileData)
+			if err != nil {
+				logger.DebugContext(ctx, "error extracting private key", "path", path, "error", err)
+			} else {
+				s.privateKeys[path] = key
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		logger.WarnContext(ctx, "error walking directory", "root", root, "error", err)
+	}
+}
+
+var (
+	supportedPrivateKeyHeaders = [][]byte{
+		[]byte("RSA PRIVATE KEY"),
+		[]byte("PRIVATE KEY"),
+		[]byte("EC PRIVATE KEY"),
+		[]byte("DSA PRIVATE KEY"),
+		[]byte("OPENSSH PRIVATE KEY"),
+	}
+)
+
+// readFileIfSSHPrivateKey checks if a file is an OpenSSH private key
+func readFileIfSSHPrivateKey(filePath string) ([]byte, bool, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, false, err
+	}
+	defer file.Close()
+
+	// read the first 150 bytes of the file to check if it's an OpenSSH private key.
+	var buf [150]byte
+	n, err := file.Read(buf[:])
+	if errors.Is(err, io.EOF) || n < len(buf) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, trace.Wrap(err, "failed to read file")
+	}
+
+	isPrivateKey := false
+	for _, header := range supportedPrivateKeyHeaders {
+		if bytes.Contains(buf[:], header) {
+			isPrivateKey = true
+			break
+		}
+	}
+	if !isPrivateKey {
+		return nil, false, nil
+	}
+
+	// read the entire file
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return nil, false, trace.Wrap(err, "failed to read file")
+	}
+	return append(buf[:], data...), true, nil
+}
+
+func extractSSHKey(ctx context.Context, path, deviceID string, fileData []byte) (*accessgraphsecretsv1pb.PrivateKey, error) {
+	logger := slog.Default().With("private_key_file", path, "device_id", deviceID)
+
+	var publicKey ssh.PublicKey
+	var mode accessgraphsecretsv1pb.PublicKeyMode
+	var pme *ssh.PassphraseMissingError
+	switch pk, err := ssh.ParsePrivateKey(fileData); {
+	case errors.As(err, &pme):
+		if pme.PublicKey != nil {
+			// If the key is a OpenSSH private key whose public key is embedded in the header, it will return the public key. This is
+			// a special case for OpenSSH private keys that have the public key embedded in the header, for more information see
+			// OpenSSH's ssh key format: https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.key
+			publicKey = pme.PublicKey
+			mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED
+			break
+		}
+		publicKey, mode = parsePublicKeyFromPublicPath(ctx, logger, path)
```

Comment:
Suggestion: It's strange that this function doesn't return an error. I would keep the mode=PROTECTED assumption outside of it.

Eg:

```go
		publicKey, err = parsePublicKeyFromPublicPath(ctx, logger, path+".pub")
		if err != nil {
			mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED
			logger.DebugContext(ctx, "Unable to read public key file", "err", err)
		} else {
			mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PUB_FILE
		}
```

Reply:
You do get errors in the function, it just happens that they are automatically swallowed. Erroring "normally" and let the caller decide to swallow is a design choice you could make. (It's only a suggestion, but trying to clarify the intent of the comment.)